### PR TITLE
Feat/4 first tasks

### PR DESCRIPTION
**Pull request by kravaring**
#### 🤔 This is a ...

- [ ] 🌟 New task
- [x] 🌐 New module
- [ ] ⚙️ Update to an existing task
- [ ] 🔧 Update to an existing module
- [ ] 🔗 Update or addition of external resources or links
- [ ] 🐛 Fix in a task or related content
- [ ] 🛠 Fix in a module or related content
- [ ] ✏️ Fixed a typo or grammatical error
- [ ] 🔗 Fixed a broken link
- [ ] ❓ Other (specify: **\*\*\*\***\_\_\_\_**\*\*\*\***)

#### Description

- **Brief Overview:**
  
- **Implementation Approach:**
  

#### Additional Information

- **Screenshots/Links:**
  
- [ ] Related Issues:

#### Checklist

- [x] ✅ I have performed a self-review of my own code.
- [x] 📝 I have commented my code, particularly in hard-to-understand areas.
- [x] 🔧 I have made corresponding changes to the documentation (if applicable).
- [x] 🚫 My changes generate no new warnings or errors.

### CODE REVIEW

File: react/README.md [hunk 1]
```diff
@@ -17,29 +17,55 @@ This course is aimed to the students of the RS School which passed RS School sta
...
-### Week #1 (23th of October, 2023)
+### Week #1 (25th of June, 2024)
 
-[React. Components](modules/module01)
+#### Theory
 
-### Week #2 (30th of October, 2023)
+[Vite with React and the TypeScript](modules/react-setup-env/)
+[Class Components](modules/class-component/)
+[Error Boundaries](modules/error-boundary/)
```
Comment (andron13) [inline react/README.md:26]: Bulleted list ор enumeration is missing for the items.

File: react/README.md [hunk 2]
```diff
@@ -17,29 +17,55 @@ This course is aimed to the students of the RS School which passed RS School sta
...
-[React. Tests / Context API](modules/module03)
+### Week #2 (2nd of July, 2024)
 
-### Week #4 (13th of November, 2023)
+#### Theory
 
-[React. Redux. RTK](modules/module04)
+[Hooks](modules/hooks/)
+[React Router](modules/router/)
+[React Tests](modules/testing/)
```
Comment (andron13) [inline react/README.md:38]: Bulleted list ор enumeration is missing for the items.

File: react/README.md [hunk 3]
```diff
@@ -17,29 +17,55 @@ This course is aimed to the students of the RS School which passed RS School sta
...
 
-### Week #6 (27th of November, 2023)
+### Week #3 (9th of July, 2024)
 
-[React. Forms](modules/module06)
+#### Theory
+
+[Context API](modules/context-api/)
+[Redux. RTK. RTK Query](modules/react-redux/)
+
```
Comment (andron13) [inline react/README.md:50]: Bulleted list ор enumeration is missing for the items.

File: react/README.md [hunk 4]
```diff
@@ -17,29 +17,55 @@ This course is aimed to the students of the RS School which passed RS School sta
...
+#### Practical tasks
+
+[Task 3](modules/tasks/redux.md)
+
+### Week #4 (16th of July, 2024)
+
+#### Theory
+
+[Context API](modules/context-api/)
+[Redux. RTK. RTK Query](modules/react-redux/)
```
Comment (andron13) [inline react/README.md:60]: Bulleted list ор enumeration is missing for the items.

**Review by andron13 — COMMENTED** [review]
Found a couple of shortcomings.

File: react/modules/tasks/redux.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# Redux. Redux Toolkit, RTK Query. Context api.
+
+## Technical requirements.
+
+1. Create a separate branch for this task from the previous branch task.
+2. Redux Integration:
+   - Integrate Redux into your application. You'll need to set up the Redux store and reducers using Redux Toolkit.
+3. RTK Query integration:
+   - RTK Query Implementation: Use Redux Toolkit Query (RTK Query) to make API calls and cache the results. This will modify your previous API call implementation;
+   - Use RTK Query to show whether the data is being loaded. Avoid prop-drilling, if it is required to show indicator in child component, put the respective flag to the store;
```
Comment (SpaNb4) [inline react/modules/tasks/redux.md:10]: if it is required to show an indicator in a child component, put the respective flag in the store;

File: react/modules/tasks/redux.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+   - When user unselects an item, it should be removed from the store;
+2. Downloading selected items (Redux):
+   - When at least 1 item has been selected, the flyout element should appear on the bottom;
+   - The flyout element should contain number of selected elements (e.g. "3 items are selected") and 2 buttons: "Unselect all" and "Download";
+   - When "Unselect all" button is clicked, all the selected items should be unselected and the flyout should be removed from the page;
+   - When "Download" button is clicked, you should save the list of selected items (e.g. name, description, details url, any other useful information) to the .csv file, name should contain the number of selected items (e.g. if you've selected "episode" endpoint of the Star Trek api and selected 15 items, the name might be "15_episodes.csv")
+3. Custom theme (Context API):
+   - Add button/radio buttons/dropdown on top of the application for theme selection;
+   - User should have an option to select one of the 2 themes (e.g. light or dark);
+   - Selected theme should affect the appearence of the whole application
```
Comment (SpaNb4) [inline react/modules/tasks/redux.md:33]: appearance

File: react/modules/tasks/redux.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+   - Test the functionality related to Redux state and API calls.
+
+## Application Requirements.
+
+1. Selected items (Redux):
+   - Each item on the dashboard should have a checkbox, information about which items have been selected should be stored in the store;
+   - When user navigates to the next page, and then goes back, previously selected items should be shown (if there were any);
+   - When user unselects an item, it should be removed from the store;
+2. Downloading selected items (Redux):
+   - When at least 1 item has been selected, the flyout element should appear on the bottom;
```
Comment (SpaNb4) [inline react/modules/tasks/redux.md:26]: at the bottom

File: react/modules/tasks/class-components.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
+# React project setup. Class components. Error boundary.
+
+## Technical Requirements.
```
Comment (SpaNb4) [inline react/modules/tasks/class-components.md:3]: Personally, I don't like those periods at the end

File: react/modules/tasks/routing.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+1. Create a separate branch for this task from the previous task's branch.
+2. All components must be changed to **functional components**, except **Error Boundary** components, as error boundaries in React still need to be class components.
+   - Implement custom hook to restore search query from the local storage (LS) and put it to the LS on unmount. Use respective React lifecycle hook as a basis.
+3. All logic should be split into components:
+   - If you need an access either to the component's lifecycle or the state **use hooks**.
+   - All data should be stored in the **component's state**.
+4. Add routing to your application using **React Router**.
+5. Add 404 page when user navigates to non-existing route.
+6. Add and configure test runner: Jest or Vitest. Test runner should show the test coverage. You should aim to reach at least 80% of the test coverage.
+7. Add testing library: React Testing Library. You should add tests for the several scenarios keeping in mind that mocked data should be used instead of real API calls.
```
Comment (SpaNb4) [inline react/modules/tasks/routing.md:14]: Add a 404 page
Add and configure a test runner
Add a testing library

File: react/modules/tasks/routing.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+   - right section should display details using Router Outlet (show loading indicator while making an additional call for details, add control for closing the section, also section should be closed when user clicks on the left section)
+   - reflect in the url that "Details" section has been opened for the selected item (e.g. /?frontpage=2&details=1).
+
+## Points.
+
+A student can achieve a maximum of 100 points.
+
+### Cross-check (score can be less if some parts of the functionality don't work)
+
+- Tests are added to the project, can be run via a sepcial command in the pacakge.json, no errors are produced, test coverage >= 80% - **30**
```
Comment (SpaNb4) [inline react/modules/tasks/routing.md:56]: via a special command in the package.json

Comment (SpaNb4) [reply react/modules/tasks/routing.md:56]: We should explicitly specify the column that students should check, e.g. smthg

File: react/modules/tasks/routing.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+- Conversion of class components to functional components with hooks - **15**
+- Custom hook to restore search query from LS and put it back to LS on umnount - **5**
```
Comment (SpaNb4) [inline react/modules/tasks/routing.md:58]: on unmount

File: react/modules/tasks/class-components.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+1. Create a separate branch for this task.
+2. Language Requirement
+   - Use **TypeScript** for the project.
+3. Project Setup
+   - Initialize the project using [Vite](https://vitejs.dev/guide/) with the [_react-ts_ template](https://vite.new/react-ts).
+4. Code Quality Tools
+   1. ESLint
+      - Set up ESLint to throw errors if TypeScript's _any_ type is used.
+      - Follow the [configuration guide](https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/module01/configs.md).
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/class-components.md:13]: I suggest removing the configuration for now. It is outdated anyway.

File: react/modules/tasks/routing.md (hunk 9 above)
Comment (kravaring) [reply react/modules/tasks/routing.md]: done

File: react/modules/tasks/class-components.md (hunk 8 above)
Comment (kravaring) [reply react/modules/tasks/class-components.md]: Updated

File: react/modules/tasks/redux.md (hunk 7 above)
Comment (kravaring) [reply react/modules/tasks/redux.md]: done

File: react/modules/tasks/redux.md (hunk 6 above)
Comment (kravaring) [reply react/modules/tasks/redux.md]: done

File: react/modules/tasks/redux.md (hunk 5 above)
Comment (kravaring) [reply react/modules/tasks/redux.md]: done

File: react/modules/tasks/class-components.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        - `lint`: For running the lint command.
+        - `format:fix`: For running Prettier's --write command.
+5. Pick a RESTfull api which supports search and pagination (pagination might be reffered as _offset_ and _limit_ params). E.g. https://pokeapi.co/, for Star Wars fans https://swapi.dev/api, for Star Trek fans https://stapi.co/api-documentation (OpenApi spec can be checked here https://editor.swagger.io/?url=https://stapi.co/api/v1/rest/common/download/stapi.yaml), or you can select another one complying with the requirements.
+
+## Application Requirements.
+
+1. Create a page comprised of 2 horizontal section (a smaller top one, and a bigger bottom one).
+2. On the top section put _Search_ input and the "Search" button. _Search_ component should look for a previously saved search term in the local storage (LS), if there isn't any - leave the input empty.
+3. Bottom section should show be used for displaying search results (name and a small description).
+4. By default application makes a call to the selected api to get the list of the items with the search term fron the input (only first page). If the input is empty make a call to get all the items.
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/class-components.md]: done

File: react/modules/tasks/class-components.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      - Add Husky and configure it to run linting on pre-commit.
+   4. package.json commands
+      - Add the following npm scripts:
+        - `lint`: For running the lint command.
+        - `format:fix`: For running Prettier's --write command.
+5. Pick a RESTfull api which supports search and pagination (pagination might be reffered as _offset_ and _limit_ params). E.g. https://pokeapi.co/, for Star Wars fans https://swapi.dev/api, for Star Trek fans https://stapi.co/api-documentation (OpenApi spec can be checked here https://editor.swagger.io/?url=https://stapi.co/api/v1/rest/common/download/stapi.yaml), or you can select another one complying with the requirements.
+
+## Application Requirements.
+
+1. Create a page comprised of 2 horizontal section (a smaller top one, and a bigger bottom one).
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/class-components.md]: done

File: react/modules/tasks/class-components.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+2. On the top section put _Search_ input and the "Search" button. _Search_ component should look for a previously saved search term in the local storage (LS), if there isn't any - leave the input empty.
+3. Bottom section should show be used for displaying search results (name and a small description).
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/class-components.md]: done

File: react/modules/tasks/class-components.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
+# React project setup. Class components. Error boundary.
+
+## Technical Requirements.
+
+1. Create a separate branch for this task.
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/class-components.md]: done

File: react/modules/tasks/README.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+- Avoid unnecessary repetition: DRY
+- Delete what is not needed: YAGNI
+
+We also need to mention the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) and other [SOLID](https://en.wikipedia.org/wiki/SOLID) principles
+Please, check [this article](https://dmitripavlutin.com/7-architectural-attributes-of-a-reliable-react-component/) for reference
+Last, but not least - check the presence of the comments. Ideally there shouldn't be any comments at all. Sometimes people just comment code which is not needed. So why not to remove it entirely? In case you will need to restore this code, you can always refer to the git history. And more - [comments are lies](https://blog.devgenius.io/code-should-be-the-one-version-of-the-truth-dont-add-comments-b0bcd8631a9a)
+
+## Cross-check process
+
+Run app and check that the functionality is working (cross-check)
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/README.md]: Thanks!

File: react/modules/tasks/README.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+**Do not merge Pull Request from the development branch to the `main` branch**
+
+## Cross-code-review process
+
+1. Clone the repository you are going to review
+2. Install all the required dependencies
+3. Run linting using special command in package.json file, output should not produce any errors or warnings
+4. Run tests using special command in package.json file, all tests should pass, test coverage should be shown after running all the tests
+5. Review the code. Pay attention at the following "code smells": props drilling; large, complex components aka "god" components; direct DOM manipulation, etc.
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/README.md]: Updated

File: react/modules/tasks/README.md [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+We will try to conduct a session for each module providing answers for some questions.
+
+## Repository requirements
+
+- The task should be completed in **your personal private repository**.
+- In the repository, create a branch from the `main` branch using the task's name and work within this newly created branch.
+- The commit history should reflect the process of creating the app. Refer to [Commit requirements](https://docs.rs.school/#/git-convention?id=%D0%A2%D1%80%D0%B5%D0%B1%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D0%BA-%D0%B8%D0%BC%D0%B5%D0%BD%D0%B0%D0%BC-%D0%BA%D0%BE%D0%BC%D0%BC%D0%B8%D1%82%D0%BE%D0%B2) [RU] for details.
+- Once development is complete, you need to create a Pull Request from the current task branch to the `main` branch. Refer to [Pull Request requirements](https://docs.rs.school/#/pull-request-review-process?id=%D0%A2%D1%80%D0%B5%D0%B1%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D0%BA-pull-request-pr) [RU] for details.
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/README.md]: Thanks!

File: react/modules/tasks/README.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
+# General requirements for all the tasks
+
+Those requireminets are applied to all the tasks, if not provided in the task explicitly
+
+## Questions
+
+You should be using Discord as the main mean of the communication.
+Also we will try to collect your questions regarding the 1st Module using special form, which will be provided via the Discord with the 1st Module start. Questions will be collected in Module 01 section spreadsheet, which also will be provided via the Discord with the 1st Module start. Please, check answers carefully before posting the question, may be your question has been answered already.
```
[the remark this answers is not in the document]
Comment (kravaring) [reply react/modules/tasks/README.md]: Yep

File: react/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kravaring) [reply react/README.md]: Added bullet points

File: react/README.md (hunk 4 above)
Comment (kravaring) [reply react/README.md]: Done

File: react/README.md (hunk 3 above)
Comment (kravaring) [reply react/README.md]: Done

File: react/README.md (hunk 2 above)
Comment (kravaring) [reply react/README.md]: Done

File: react/README.md (hunk 1 above)
Comment (kravaring) [reply react/README.md]: Done

File: react/modules/tasks/routing.md (hunk 10 above)
Comment (SpaNb4) [reply react/modules/tasks/routing.md]: @kravaring What're your thoughts on my second comment?

Comment (kravaring) [reply react/modules/tasks/routing.md]: @SpaNb4 , sorry, missed that part. Updated.

**Review by SpaNb4 — APPROVED** [review]

**Pull request merged**